File: General-DSA-Questions/Stack/8_sumOfSubArrayMinimum.cpp
// Given an array of integers arr, return the sum of the minimum value of all subarrays of arr.
// Since the answer can be large, return the result modulo 10^9 + 7.

// Notes link : https://takeuforward.org/data-structure/sum-of-subarray-minimums
// Video link : https://www.youtube.com/watch?v=v0e8p9JCgRc

// Example :
// Input: [3, 1, 2, 4]
// Output: 17

// All subarrays and their minimums:
// [3] => 3  
// [3,1] => 1  
// [3,1,2] => 1  
// [3,1,2,4] => 1  
// [1] => 1  
// [1,2] => 1  
// [1,2,4] => 1  
// [2] => 2  
// [2,4] => 2  
// [4] => 4  

// Sum = 3+1+1+1+1+1+1+2+2+4 = 17

// Intiution:
// Each element contributes to the sum only when it is the minimum of a subarray.
// So instead of computing all subarrays →
// We compute how many subarrays have arr[i] as their minimum.
// If we know:
//    How many subarrays start before or at i,
//    And how many subarrays end after or at i,
//    Then their product gives us the total number of subarrays where arr[i] is the minimum.

// For every index i, we find:
//    PLE[i]: index of Previous Less Element → closest element to the left that is less than arr[i].
//    NLE[i]: index of Next Less Element → closest element to the right that is strictly less than arr[i].

// Any subarray that has:
//    its left boundary between PLE[i] + 1 and i,
//    its right boundary between i and NLE[i] - 1, will have arr[i] as its minimum element.

// So, the number of such subarrays is:
//    leftCount = i - PLE[i]
//    rightCount = NLE[i] - i
//    totalSubarraysWithArrIAsMin = leftCount * rightCount

// Therefore, the contribution of arr[i] is:
// arr[i] * leftCount * rightCount


// Why totalSubarraysWithArrIAsMin = leftCount * rightCount?
// Now, think about pairing the subarrays that end at i with the subarrays that start at i.
// ->  Each subarray that ends at i with arr[i] as the minimum (from the left side) can be combined with each subarray that starts at i with arr[i] as the minimum (from the right side).
// ->  The key point here is that each subarray ending at i and having arr[i] as the minimum can be paired with each subarray starting at i and having arr[i] as the minimum.
//     This results in a "rectangular" relationship between subarrays from the left side and the right side.
// For example, if there are left ways to pick subarrays that end at i and right ways to pick subarrays that start at i, then there are left × right subarrays that both include arr[i] as the minimum.

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

const int MOD = 1e9 + 7;

int sumSubarrayMins(vector<int>& arr) {
    int n = arr.size();
    vector<int> ple(n), nle(n); // Previous Less Element and Next Less Element
    // ple[i] will store the index of the previous less element for arr[i]
    // nle[i] will store the index of the next less element for arr[i]
    // If no such element exists, ple[i] will be -1 and nle[i] will be n (size of the array)
    stack<int> s;

    // Previous Less Element (PLE)
    for (int i = 0; i < n; ++i) {
        while (!s.empty() && arr[s.top()] > arr[i]) { // Pop elements that are greater than the current element
            s.pop();
        }
        // -1 is used because, If no smaller element exists, that means all subarrays starting from index 0 up to i can have arr[i] as the minimum
        ple[i] = s.empty() ? -1 : s.top();  // If stack is empty, ple[i] is -1, otherwise it is the index of the previous less element
        // Push current index onto the stack
        s.push(i);
    }

    // Clear the stack to reuse
    while (!s.empty()) s.pop();

    // Next Less Element (NLE)
    for (int i = n - 1; i >= 0; --i) {
        while (!s.empty() && arr[s.top()] >= arr[i]) {  // Pop elements that are greater than or equal to the current element
            s.pop();
        }
        // n is used because, If no smaller element exists, that means all subarrays ending from index i to n-1 can have arr[i] as the minimum
        nle[i] = s.empty() ? n : s.top(); // If stack is empty, nle[i] is n, otherwise it is the index of the next less element
        s.push(i);
    }

    // Calculate total sum
    long long sum = 0;
    for (int i = 0; i < n; ++i) {
        long long left = i - ple[i];
        long long right = nle[i] - i;
        sum = sum + (arr[i] * left * right) % MOD; // Calculate contribution of arr[i]
        sum %= MOD; // Take modulo to avoid overflow
    }

    return (int)sum;
}

int main() {
    vector<int> arr = {3, 1, 2, 4};
    cout << "Sum of Subarray Minimums: " << sumSubarrayMins(arr) << "\n";
    return 0;
}
// Time Complexity: O(n) - Each element is pushed and popped from the stack at most once.
// Space Complexity: O(n) - We use two arrays for PLE and NLE, and a stack for processing.

File: JS-DSA-Questions/Stack/8_sumOfSubArrayMinimum.js
// Given an array of integers arr, return the sum of the minimum value of all subarrays.
// Since the answer can be large, return it modulo 1e9 + 7.

// Notes link : https://takeuforward.org/data-structure/sum-of-subarray-minimums
// Video link : https://www.youtube.com/watch?v=v0e8p9JCgRc

// Example: [3, 1, 2, 4] → 17

// Intuition:
// Each element contributes only when it is the minimum of a subarray. Instead of enumerating all
// subarrays, count how many subarrays have arr[i] as their minimum.
// For each index i, find:
//    PLE[i]: index of the Previous Less Element (closest smaller element to the left).
//    NLE[i]: index of the Next Less Element (closest strictly-smaller element to the right).
// Then:
//    leftCount  = i - PLE[i]
//    rightCount = NLE[i] - i
//    contribution = arr[i] * leftCount * rightCount
// Left and right choices multiply because any left extension pairs with any right extension.

// JavaScript note: C++ used `long long` and a MOD constant to avoid overflow. JS numbers are safe
// up to 2^53; for typical inputs plain arithmetic with % MOD is fine. For very large arr[i]*left*
// right that could exceed 2^53, BigInt would be needed, but we keep it simple here.
const MOD = 1e9 + 7;

function sumSubarrayMins(arr) {
  const n = arr.length;
  const ple = new Array(n); // previous less element indices
  const nle = new Array(n); // next less element indices
  const s = [];

  // Previous Less Element (PLE)
  for (let i = 0; i < n; i++) {
    while (s.length > 0 && arr[s[s.length - 1]] > arr[i]) {
      s.pop();
    }
    ple[i] = s.length === 0 ? -1 : s[s.length - 1];
    s.push(i);
  }

  // Clear the stack to reuse
  s.length = 0;

  // Next Less Element (NLE)
  for (let i = n - 1; i >= 0; i--) {
    while (s.length > 0 && arr[s[s.length - 1]] >= arr[i]) {
      s.pop();
    }
    nle[i] = s.length === 0 ? n : s[s.length - 1];
    s.push(i);
  }

  // Total sum
  let sum = 0;
  for (let i = 0; i < n; i++) {
    const left = i - ple[i];
    const right = nle[i] - i;
    sum = (sum + ((arr[i] * left * right) % MOD)) % MOD;
  }

  return sum;
}

console.log("Sum of Subarray Minimums:", sumSubarrayMins([3, 1, 2, 4])); // 17

// Time Complexity: O(n). Space Complexity: O(n).
